typing/parmatch.ml

```diff
@@ -137,6 +137,29 @@ let get_type_path ty tenv =
 open Format
 ;;
 
+(** Compute which labels are defined within a record pattern,
+    and report with [err] if a label is multiply defined.
+    The function returns None if the label list is empty *)
+let defined_labels ~err lbl_pat_list =
+  match lbl_pat_list with
+  | [] -> None (* should not happen, the caller should handle this *)
+  | (_, label1, _) :: _ ->
+      let all = label1.lbl_all in
+      let defined = Array.make (Array.length all) false in
+      let check_defined (_, label, _) =
+        if defined.(label.lbl_pos) then
+             err label.lbl_name
+        else defined.(label.lbl_pos) <- true in
+      List.iter check_defined lbl_pat_list;
+      Some(all,defined)
+
+let pretty_record_closure ppf lbl_list =
```

Comment:
I find the "closure" wording confusing (I think of function closures). You could maybe use "ending" instead?